File: mojo/edk/system/android/java/src/org/chromium/mojo/edk/ParcelableChannelServer.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.mojo.edk;

import android.os.IBinder;
import android.os.Parcelable;
import android.os.RemoteException;

import org.chromium.base.Log;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;

/**
 * Provides an implementation of IParcelableChannel that stores the received parcelables and make
 * them accessible to native code.
 */
@JNINamespace("mojo::edk")
public class ParcelableChannelServer extends IParcelableChannel.Stub {
    private static final String TAG = "ParcelableChanServer";

    private long mNativeParcelableChannelServer;

    private ParcelableChannelServer(long nativeParcelableChannelServer) {
        mNativeParcelableChannelServer = nativeParcelableChannelServer;
    }

    @Override
    public void sendParcelable(ParcelableWrapper parcelableValue) {
        if (mNativeParcelableChannelServer != 0) {
            nativeOnParcelableReceived(mNativeParcelableChannelServer, parcelableValue.getId(),
                    parcelableValue.getParcelable());
        }
    }

    @CalledByNative
    private static ParcelableChannelServer create(long nativeParcelableChannelServer) {
        return new ParcelableChannelServer(nativeParcelableChannelServer);
    }

    @CalledByNative
    private void invalidate() {
        mNativeParcelableChannelServer = 0;
    }

    @CalledByNative
    private boolean setChannelOnParent(IBinder parcelableChannelProvider) {
        try {
            IParcelableChannelProvider.Stub.asInterface(parcelableChannelProvider)
                    .sendParcelableChannel(this);
            return true;
        } catch (RemoteException re) {
            Log.e(TAG, "Failed to set ParcelableChannel on parent.", re);
            return false;
        }
    }

    private native void nativeOnParcelableReceived(
            long nativeParcelableChannelServer, int id, Parcelable parcelable);
}

File: mojo/edk/system/android/parcelable_channel_server.cc
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "mojo/edk/system/android/parcelable_channel_server.h"

#include "mojo/edk/jni/ParcelableChannelServer_jni.h"

using base::android::AttachCurrentThread;
using base::android::ScopedJavaGlobalRef;
using base::android::ScopedJavaLocalRef;
using base::android::JavaRef;

namespace mojo {
namespace edk {

ParcelableChannelServer::ParcelableChannelServer()
    : j_parcelable_channel_(Java_ParcelableChannelServer_create(
          AttachCurrentThread(),
          reinterpret_cast<intptr_t>(this))) {}

ParcelableChannelServer::~ParcelableChannelServer() {
  Java_ParcelableChannelServer_invalidate(AttachCurrentThread(),
                                          j_parcelable_channel_);
}

ParcelableChannelServer::ParcelableChannelServer(
    ParcelableChannelServer&& other) = default;

ParcelableChannelServer& ParcelableChannelServer::operator=(
    ParcelableChannelServer&& other) = default;

void ParcelableChannelServer::SetListener(Listener* listener) {
  DCHECK(!listener_);
  listener_ = listener;
}

ScopedJavaGlobalRef<jobject> ParcelableChannelServer::GetIParcelableChannel() {
  // The ParcelableChannelServer java class implements IParcelableChannel.
  return j_parcelable_channel_;
}

bool ParcelableChannelServer::SetOnChannelProvider(
    const JavaRef<jobject>& channel_provider) {
  DCHECK(!channel_provider.is_null());
  return Java_ParcelableChannelServer_setChannelOnParent(
      AttachCurrentThread(), j_parcelable_channel_, channel_provider);
}

void ParcelableChannelServer::OnParcelableReceived(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller,
    jint id,
    const base::android::JavaParamRef<jobject>& parcelable) {
  DCHECK(listener_);
  if (listener_) {
    listener_->OnParcelableReceived(id, parcelable);
  }
}

}  // namespace edk
}  // namespace mojo